feat: move vertical splitting & detect non-linear data

Joe-Blount opened the pull request:
This PR moves the vertical splitting up to the area of the `RoundInfo` calculation.  There is more knowledge of the total situation there, making us better able to make intelligent splitting decisions.  Previously the lack of context made vertical splitting somewhat crude, so it was only applied in some specific situations.  Now that its smarter about where & why it splits, we can make use of vertical splitting any time the chain of overlapping L0s is larger than the max compact size.  So the complicated criteria to decide to split is gone.

I suggest reviewing this commit by commit.
Commit 1: Changes to the some tests before making functional changes.
  - the artificially small timestamps in `actual_case_from_catalog_1` skews the write amplification.  So all timestamps are multiplied by 10,000.
  - `l0s_almost_needing_vertical_split` now gets split, so I modified it to have smaller files, and kept an unmodified version that does split.

Commit 2: Moves vertical splitting to the RoundInfo calculation time.
  - `select_split_times` is modified to now take hints, to align to target level file timestamps

Commit 3: Insta test churn from moving vertical splitting  (a widespread improvement in write amplification)

Commit 4: Add the ability to detect non-linear distribution of data
  - Vertical splitting now forces splits when the data density changes significantly, and lets `select_split_times` chose the time to split within regions of similar density (choosing times to align to target level files that must be compacted with)

Commit 5: Add tests for non-linear data distribution
  - Patterned after the simulator's display of files, it now can display file regions.
  - 6 specific scenarios are tested, using the visual display to show how the non-linear data detection identifies and handles varying data density.

Commit 6: More Insta churn (more write amplification improvement)

helps: https://github.com/influxdata/idpe/issues/17974

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

Joe-Blount commented on compactor/src/components/split_or_compact/start_level_files_to_split.rs:
```diff
@@ -438,50 +464,64 @@ pub fn identify_start_level_files_to_split(
 
         // splitting 150 bytes based on a max of 100, with a time range 0-100, gives 2 splits, into 3 pieces.
         // 1 split into 2 pieces would have also been ok.
-        let mut split_times = super::select_split_times(150, 100, 0, 100);
```
The approach to these tests is to add an empty vec of hints to the test cases we already had.  That shows `select_split_times` still behaves like it did before if no hints are given.  Then each of of the previous test cases is repeated with some hints.  If the hints are close to what `select_split_times` would have chosen on its own, we see that the hint is adopted.  If the hints are far from what `select_split_times` would have done on its own, we see the hint is not adopted.

Joe-Blount commented on compactor/tests/layouts/backfill.rs:
```diff
@@ -137,320 +137,711 @@ async fn random_backfill_empty_partition() {
-    - "L2.?[221,420] 1.05us 101mb|------------------L2.?------------------|                                                "
-    - "L2.?[421,619] 1.05us 100mb                                          |------------------L2.?------------------|      "
-    - "L2.?[620,643] 1.05us 12mb                                                                                     |L2.?|"
+    - "L1.200[677,986] 1.04us 92mb                             |--------------------------L1.200--------------------------| "
+    - "L1.206[529,676] 1.04us 78mb|----------L1.206----------|                                                              "
+    - "**** 3 Output Files (parquet_file_id not yet assigned), 248mb total:"
+    - "L1                                                                                                                 "
+    - "L1.?[529,714] 1.05us 101mb|---------------L1.?---------------|                                                      "
+    - "L1.?[715,899] 1.05us 100mb                                    |---------------L1.?---------------|                  "
+    - "L1.?[900,986] 1.05us 47mb                                                                         |-----L1.?-----| "
     - "Committing partition 1:"
-    - "  Soft Deleting 3 files: L1.97, L1.98, L1.99"
-    - "  Upgrading 1 files level to CompactionLevel::L2: L1.96"
+    - "  Soft Deleting 5 files: L1.200, L0.201, L0.202, L1.206, L0.208"
     - "  Creating 3 files"
-    - "**** Simulation run 24, type=split(CompactAndSplitOutput(TotalSizeLessThanMaxCompactSize))(split_times=[952]). 1 Input Files, 96mb total:"
-    - "L1, all files 96mb                                                                                                 "
-    - "L1.101[819,986] 1.05us   |-----------------------------------------L1.101-----------------------------------------|"
-    - "**** 2 Output Files (parquet_file_id not yet assigned), 96mb total:"
+    - "**** Simulation run 65, type=split(CompactAndSplitOutput(FoundSubsetLessThanMaxCompactSize))(split_times=[594, 783]). 3 Input Files, 262mb total:"
+    - "L1                                                                                                                 "
+    - "L1.211[405,528] 1.05us 61mb|-------L1.211-------|                                                                    "
+    - "L1.212[529,714] 1.05us 101mb                      |------------L1.212-------------|                                   "
+    - "L1.213[715,899] 1.05us 100mb                                                        |------------L1.213-------------| "
+    - "**** 3 Output Files (parquet_file_id not yet assigned), 262mb total:"
     - "L2                                                                                                                 "
-    - "L2.?[819,952] 1.05us 77mb|--------------------------------L2.?---------------------------------|                   "
-    - "L2.?[953,986] 1.05us 19mb                                                                        |-----L2.?------| "
+    - "L2.?[405,594] 1.05us 101mb|--------------L2.?--------------|                                                        "
+    - "L2.?[595,783] 1.05us 100mb                                  |--------------L2.?--------------|                      "
+    - "L2.?[784,899] 1.05us 61mb                                                                     |-------L2.?-------| "
     - "Committing partition 1:"
-    - "  Soft Deleting 1 files: L1.101"
-    - "  Upgrading 1 files level to CompactionLevel::L2: L1.100"
-    - "  Creating 2 files"
-    - "**** Final Output Files (2.89gb written)"
+    - "  Soft Deleting 3 files: L1.211, L1.212, L1.213"
+    - "  Upgrading 2 files level to CompactionLevel::L2: L1.209, L1.210"
+    - "  Creating 3 files"
+    - "**** Final Output Files (2.34gb written)"
```
write amp gets better, but an L1 file is left behind now (whether or not an L1 is left depends on some on luck - as we run out of L1s to compact, if we have less than the compaction threshold some can be left behind).

Joe-Blount commented on compactor/tests/layouts/backfill.rs:
```diff
@@ -599,435 +990,951 @@ async fn random_backfill_over_l2s() {
-    - "**** Simulation run 31, type=split(CompactAndSplitOutput(FoundSubsetLessThanMaxCompactSize))(split_times=[764]). 2 Input Files, 156mb total:"
+    - "**** Simulation run 86, type=split(CompactAndSplitOutput(FoundSubsetLessThanMaxCompactSize))(split_times=[765]). 2 Input Files, 155mb total:"
     - "L1                                                                                                                 "
-    - "L1.119[700,799] 1.05us 56mb|----------------------------------------L1.119-----------------------------------------| "
+    - "L1.243[700,799] 1.05us 55mb|----------------------------------------L1.243-----------------------------------------| "
     - "L2                                                                                                                 "
     - "L2.8[700,799] 799ns 100mb|-----------------------------------------L2.8------------------------------------------| "
-    - "**** 2 Output Files (parquet_file_id not yet assigned), 156mb total:"
+    - "**** 2 Output Files (parquet_file_id not yet assigned), 155mb total:"
     - "L2                                                                                                                 "
-    - "L2.?[700,764] 1.05us 102mb|--------------------------L2.?--------------------------|                                "
-    - "L2.?[765,799] 1.05us 55mb                                                           |------------L2.?------------| "
+    - "L2.?[700,765] 1.05us 102mb|--------------------------L2.?---------------------------|                               "
+    - "L2.?[766,799] 1.05us 53mb                                                           |-----------L2.?------------|  "
     - "Committing partition 1:"
-    - "  Soft Deleting 2 files: L2.8, L1.119"
+    - "  Soft Deleting 2 files: L2.8, L1.243"
     - "  Creating 2 files"
-    - "**** Simulation run 32, type=split(CompactAndSplitOutput(FoundSubsetLessThanMaxCompactSize))(split_times=[864]). 3 Input Files, 156mb total:"
+    - "**** Simulation run 87, type=split(CompactAndSplitOutput(FoundSubsetLessThanMaxCompactSize))(split_times=[865]). 3 Input Files, 155mb total:"
     - "L1                                                                                                                 "
-    - "L1.120[800,874] 1.05us 42mb|-----------------------------L1.120------------------------------|                       "
-    - "L1.116[875,899] 1.05us 14mb                                                                    |------L1.116-------| "
+    - "L1.244[800,877] 1.05us 43mb|-------------------------------L1.244-------------------------------|                    "
+    - "L1.240[878,899] 1.05us 12mb                                                                      |-----L1.240------| "
     - "L2                                                                                                                 "
     - "L2.9[800,899] 899ns 100mb|-----------------------------------------L2.9------------------------------------------| "
-    - "**** 2 Output Files (parquet_file_id not yet assigned), 156mb total:"
+    - "**** 2 Output Files (parquet_file_id not yet assigned), 155mb total:"
     - "L2                                                                                                                 "
-    - "L2.?[800,864] 1.05us 102mb|--------------------------L2.?--------------------------|                                "
-    - "L2.?[865,899] 1.05us 55mb                                                           |------------L2.?------------| "
+    - "L2.?[800,865] 1.05us 102mb|--------------------------L2.?---------------------------|                               "
+    - "L2.?[866,899] 1.05us 53mb                                                           |-----------L2.?------------|  "
     - "Committing partition 1:"
-    - "  Soft Deleting 3 files: L2.9, L1.116, L1.120"
+    - "  Soft Deleting 3 files: L2.9, L1.240, L1.244"
     - "  Creating 2 files"
-    - "**** Final Output Files (4.05gb written)"
+    - "**** Final Output Files (4.04gb written)"
```
minor improvement in write amp

Joe-Blount commented on compactor/tests/layouts/backfill.rs:
```diff
@@ -2863,257 +3770,2738 @@ async fn actual_case_from_catalog_1() {
+    - "L0, all files 197mb                                                                                                "
+    - "L0.?[2080000,2229999] 3.42ms|------------------------------------------L0.?------------------------------------------|"
+    - "Committing partition 1:"
+    - "  Soft Deleting 8 f
```
write amp gets about 5% worse on this one, but this scenario has unrealistically dense data because the timestamps are fake, so it ends up splitting extra.  The test case originated from prod catalog contents, but the timestamps did not.  If the timestamps were realistic, this would behave better.

Joe-Blount commented on compactor/tests/layouts/many_files.rs:
```diff
@@ -4670,17 +5725,23 @@ async fn l0s_almost_needing_vertical_split() {
-    - "**** Final Output Files (5.13gb written)"
-    - "L1                                                                                                                 "
-    - "L1.3091[92,100] 1.02us 88mb                                                                                |L1.3091| "
+    - "**** Final Output Files (4.32gb written)"
```
a nice improvement in write amplification + the L1 file is gone (so write amplification improved more than indicated by the reduction in bytes written)

wiedld reviewed (commented):
~50% done with the code review.

Nothing blocking thus far. 
~The use of `=` comparison, when really want overlapping (not just at the boundary of splits) is the most important requested change.~ We discussed this. It's about where the split counts the edge (left versus right edge) of the next target_level files.

Joe-Blount commented on compactor/src/components/file_classifier/split_based.rs:
```diff
@@ -128,73 +129,78 @@ where
-                target_level,
-            );
-        }
+                *start_level,
```
This is still getting the level from the `RoundInfo`.  I didn't see a practical difference between calling the getter function or using the structure deconstruct*.  I was thinking keep the function for where its convenient...?

*not sure if that's what its called

wiedld replied:
There is no practical difference. My question is about consistency. Whether we should (a) use the target_level() method everywhere (so a single match on enum determines it), or to (b) always use the decision made in round_info (which means to remove the target_level() method).

Joe-Blount commented on compactor/src/components/file_classifier/split_based.rs:
```diff
@@ -258,3 +264,45 @@ fn file_classification_for_many_files(
+    for f in files {
+        let this_file_splits: Vec<i64> = split_times
+            .iter()
+            .filter(|split| split >= &&f.min_time.get() && split < &&f.max_time.get())
```
(same explanation as above, split time is the last time on the `left`)

Joe-Blount commented on compactor/tests/layouts/many_files.rs:
```diff
@@ -4698,7 +4698,7 @@ async fn l0s_needing_vertical_split() {
         .with_max_num_files_per_plan(max_files_per_plan as usize)
         .with_max_desired_file_size_bytes(max_file_size)
-        .with_partition_timeout(Duration::from_millis(10000))
+        .with_partition_timeout(Duration::from_millis(100000))
```
Adding more computation time evaluating vertical splitting slows it down.

wiedld commented:
Our over-arching goal is to have more decisions in a single location. To that end, I'm hoping we'll eventually have all of the file division done in the same place.

* Right now we have:
   * RoundInfo::SimulatedLeadingEdge 
   * RoundInfo::SimulatedLeadingEdge => in round_info (`divide()`)
   * RoundInfo::ManySmallFiles =>  in round_info (`divide()`)
   * RoundInfo::VerticalSplit => in file classifier 
        * `file_classification_for_vertical_split()` method added in this PR
   * RoundInfo::TargetLevel => divides files in both places
        * in file classifier (existing code, updated here)
        * in round_info (`divide()`) -- in new code, added here.

* Action items:
   * In this PR: 
        * RoundInfo::VerticalSplit == have file division occur in round_info (`divide()`)
   * In followup PR:  
        * RoundInfo::TargetLevel == this PR introduces splitting in both locations. Can we have a followup PR that centralizes this decision?

wiedld commented on compactor/src/components/file_classifier/split_based.rs:
```diff
@@ -128,73 +129,78 @@ where
-        let (files_to_compact, mut files_to_keep) = self
-            .target_level_split
-            .apply(files_to_compact, target_level);
+            RoundInfo::TargetLevel { target_level, .. } => {
```
I think we talked past each other on the target_level thing.

Everything else gets target_level from the round info, except that RoundInfo::SimulatedLeadingEdge uses the `target_level()` method.
Can we delete this method, and have RoundInfo::SimulatedLeadingEdge also use a round_info target_level?

Joe-Blount commented:
> Our over-arching goal is to have more decisions in a single location. To that end, I noticed a pattern in where our file division happens. What do you think, should we also be having file division occur in the same place?
> 
> * Right now we have:
>   
>   * RoundInfo::SimulatedLeadingEdge => in round_info (`divide()`)
>   * RoundInfo::ManySmallFiles =>  in round_info (`divide()`)
>   * RoundInfo::VerticalSplit => in file classifier
>     
>     * `file_classification_for_vertical_split()` method added in this PR
>   * RoundInfo::TargetLevel => divides files in both places
>     
>     * in file classifier (existing code, updated here)
>     * in round_info (`divide()`) -- in new code, added here.
> * Proposed action items?
>   
>   * In this PR:
>     
>     * RoundInfo::VerticalSplit == have file division occur in round_info (`divide()`)?
>   * In followup PR:
>     
>     * RoundInfo::TargetLevel == this PR introduces splitting in both locations. Can we have a followup PR that centralizes this decision? (TargetLevel is a special beast, is the default, and is still very split brain.)


You raise a good point about there being some confusion here.  I would identify 3 activities & locations of interest to this discussion:

RoundSplit::split separates the files we need to be working on in this round, from the rest.  Currently, all the RoundInfo types have simple criteria for this (and hopefully will remain simple).  

DivideInitial::divide groups the files for this round into branches that can be concurrently operated on. I think the name should be changed to something about branches, but I've been ignoring that for now. vertical splitting doesn't have much here because every split can be in the same branch.

classify creates the FileClassification object, which distinguishes between files that will be upgraded, compacted or split.  

That has been confusing and bugging me, but I'm not sure yet what fits long term.  I think my description above fits what they do, and may continue to be fit long term.  But there is quite a bit more change coming.  For example, I suspect ManySmallFiles will go away.  And additional RoundInfo types will be added. I'm not sure yet how well the new stuff will fit the descriptions above, but :crossed_fingers:.  I've been trying to sort out what place (if any) each of these file set manipulation layers should have, but I'm hoping to postpone cleaning them up until the bigger changes progress some more and (hopefully) the suitable role for these layers becomes more clear.

How about for now I add a comment block above each of these 3 describing what I think their responsibility is, and see how that fits through the next few PRs?  Then after more things settle out, I can clean them up as necessary.

I added this checklist to https://github.com/influxdata/idpe/issues/17974
- look over the use of getters vs. deconstructing structures to get struct contents from RoundInfo
- audit SplitReason enum usage, remove unused values
- revisit RoundSplit::split, DivideInitial::divide, and classify; what is their actual responsibilities? Is there redundancy or overlap that should be cleaned up? Is there better naming that can convey their job?

wiedld commented:
> How about for now I add a comment block above each of these 3 describing what I think their responsibility is, and see how that fits through the next few PRs? Then after more things settle out, I can clean them up as necessary.
>
> I added this checklist to https://github.com/influxdata/idpe/issues/17974

Thank you!  ❤️

wiedld reviewed (commented):
The formation of the capacity distribution is lovely 😻 .

The decisions made during distribution consolidation, are confusing. Can we chat about this on Monday?

wiedld commented on compactor/src/components/split_or_compact/start_level_files_to_split.rs:
```diff
@@ -645,4 +646,403 @@ mod tests {
+        ---
+        - case 3 linear data distribution ranges
+        - "[100,100] 100mb          |range|                                                                                   "
+        - "[101,320099] 4.43gb      |-------------------------------range--------------------------------|                    "
```
Ok, I re-thinking this -- I believe your objective is to provide a split hint whenever there is a change in density. So It's about producing split_hints matching the timepoint when density changes, and NOT about providing a split_hint after X mb are seen.

If this is the case, then yes this code change fulfills that objective.

Joe-Blount replied:
yes, this is about changes in data density.  And the linear distributions aren't just hints - the files will be split at exactly those points.  So we don't want to go overboard with lots of splitting from linear distributions.  Its just trying to isolate the big changes in density.

Joe-Blount commented on compactor/src/components/split_or_compact/start_level_files_to_split.rs:
```diff
@@ -645,4 +646,403 @@ mod tests {
+    "###
+        );
+
+        // Case 2: 1 file, even when its 1/10th the max compact size is still a single region because its consistent density.
```
yes - good catch.

wiedld reviewed (approved):
TY for making the changes! ❤️
